### ISSUE TITLE
Analyse and fix Startup/Timing problems

### ISSUE THREAD

**Issue by BenediktNiehues** [marinmitev/smarthome#85]
The IntegrationTest "assert that a rule added by json is executed correctly" only sometimes gets green when executed in eclipse, but never when executed in maven. There seems to be a timing problem on startup. This has to be analyzed

**Comment by marinmitev** [comment]
I have solved the errors in the integration test - the issues found were:
1. rules from json files are imported only once and when first test is finished these rules are disposed, could be solved by using single static instance of VolatileStorageService but this causes issues with the ItemRegistry. Solution was to move the json test in separate groovy test file which is executed first
2. small fixes in the rule engine core implementation - mainly NPE
3. improved debug messages and fixed typos
4. fix in BasicModuleHandlerFactory when creating EventConditionHandler
5. Event delivery issue when changing Item state - sometimes the Item changes its state after the GenericEventTrigger which triggers the rule and the rule condition checks the old item state - seems to be a general issue in the Event Manager and event delivery order

**Comment by marinmitev** [comment]
see TODO comment on line 177 in AutomationIntegrationTest2
https://github.com/marinmitev/smarthome/commit/fca8bf3dc71526e1a6722029a1e6a341e5ae85cd#diff-40b59eeaacedd9761988c0e6c62c1055R177

**Comment by kaikreuzer** [comment]
@BenediktNiehues Could you have a look at 1 and 6?

**Comment by marinmitev** [comment]
Here is a sample test code to check the event handling

``` groovy
        def eventHandler = [
            receive: { Event e ->
                logger.info("Event1: " + e.topic + " = " + e)
                logger.warn('after event1, myMotionItem.state=' + myMotionItem.getState());
            },
            getSubscribedEventTypes: {
                Sets.newHashSet(ItemStateEvent.TYPE)
            },
            getEventFilter:{ null }
        ] as EventSubscriber
        registerService(eventHandler)

///////////////////////
        def eventHandler2 = [
            receive: { Event e ->
                logger.info("Event2: " + e.topic + " = " + e)
                logger.warn('after event2, myMotionItem.state=' + myMotionItem.getState());
            },
            getSubscribedEventTypes: {
                Sets.newHashSet(ItemStateEvent.TYPE)
            },
            getEventFilter:{ null }
        ] as EventSubscriber
        registerService(eventHandler2)
///////////////////////
        def eventHandler3 = [
            receive: { Event e ->
                logger.info("Event3: " + e.topic + " = " + e)
                logger.warn('after event3, myMotionItem.state=' + myMotionItem.getState());
            },
            getSubscribedEventTypes: {
                Sets.newHashSet(ItemStateEvent.TYPE)
            },
            getEventFilter:{ null }
        ] as EventSubscriber
        registerService(eventHandler3)
///////////////////////
```

Then when event is posted 
        eventPublisher.post(ItemEventFactory.createStateEvent("myMotionItem", OnOffType.ON))
the messages printed are:
- Event2: smarthome/items/myMotionItem/state = myMotionItem updated to ON
- after event2, myMotionItem.state=NULL
- Event1: smarthome/items/myMotionItem/state = myMotionItem updated to ON
- after event1, myMotionItem.state=ON
- Event3: smarthome/items/myMotionItem/state = myMotionItem updated to ON
- after event3, myMotionItem.state=ON

**Comment by hristostanchev** [comment]
Benedikt need to recheck it and issue can be closed.

**Comment by BenediktNiehues** [comment]
1. rules from json files are imported only once and when first test is finished these rules are disposed, could be solved by using single static instance of VolatileStorageService but this causes issues with the ItemRegistry. Solution was to move the json test in separate groovy test file which is executed first
   --> This did not work for me. Maybe we have to create an extra bundle for it to separate it fully from the other tests.
2. Event delivery issue when changing Item state - sometimes the Item changes its state after the GenericEventTrigger which triggers the rule and the rule condition checks the old item state - seems to be a general issue in the Event Manager and event delivery order
   --> This in deed could be a problem if the state of the item is checked before it was updated. Such cases should be handled with in/outputs then. But this has to be documented very well...

I am going to try out these points on Monday. All other points were clear and ok. Thanks!

**Comment by BenediktNiehues** [comment]
I created PR #98 to make the tests more stable. Running in eclipse always works fine, but when running the "AutomationIntegrationJsonTest" (renamed from AutomationIntegrationTest2) with maven, the test did not pass because the rule is not executed. It seems that the rule with it's modules is not working correctly, although it gets the state idle. But in the logs you can see, that the ConnectionValidator does not pass...

```
Running org.eclipse.smarthome.automation.integration.test.AutomationIntegrationJsonTest
java.lang.IllegalArgumentException: Condition Type "EventCondition" does not exist!
    at org.eclipse.smarthome.automation.core.util.ConnectionValidator.validateConditionConnections(ConnectionValidator.java:201)
    at org.eclipse.smarthome.automation.core.util.ConnectionValidator.validateConnections(ConnectionValidator.java:73)
    at org.eclipse.smarthome.automation.core.util.ConnectionValidator.validateConnections(ConnectionValidator.java:56)
    at org.eclipse.smarthome.automation.internal.core.RuleEngine.setRule(RuleEngine.java:312)
    at org.eclipse.smarthome.automation.internal.core.RuleRegistryImpl.addIntoRuleEngine(RuleRegistryImpl.java:87)
    at org.eclipse.smarthome.automation.internal.core.RuleRegistryImpl.add(RuleRegistryImpl.java:77)
    at org.eclipse.smarthome.automation.internal.core.RuleRegistryImpl.add(RuleRegistryImpl.java:1)
    at org.eclipse.smarthome.automation.internal.core.provider.RuleResourceBundleImporter.importData(RuleResourceBundleImporter.java:209)
    at org.eclipse.smarthome.automation.internal.core.provider.RuleResourceBundleImporter.processAutomationProvider(RuleResourceBundleImporter.java:183)
    at org.eclipse.smarthome.automation.internal.core.provider.AutomationResourceBundlesEventQueue.processBundleChanged(AutomationResourceBundlesEventQueue.java:319)
    at org.eclipse.smarthome.automation.internal.core.provider.AutomationResourceBundlesEventQueue.run(AutomationResourceBundlesEventQueue.java:144)
    at java.lang.Thread.run(Thread.java:744)
java.lang.IllegalArgumentException: Condition Type "EventCondition" does not exist!
    at org.eclipse.smarthome.automation.core.util.ConnectionValidator.validateConditionConnections(ConnectionValidator.java:201)
    at org.eclipse.smarthome.automation.core.util.ConnectionValidator.validateConnections(ConnectionValidator.java:73)
    at org.eclipse.smarthome.automation.core.util.ConnectionValidator.validateConnections(ConnectionValidator.java:56)
    at org.eclipse.smarthome.automation.internal.core.RuleEngine.setRule(RuleEngine.java:312)
    at org.eclipse.smarthome.automation.internal.core.RuleEngine.moduleTypeUpdated(RuleEngine.java:1061)
    at org.eclipse.smarthome.automation.internal.core.type.ModuleTypeManager.addingService(ModuleTypeManager.java:217)
    at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:941)
    at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:1)
    at org.osgi.util.tracker.AbstractTracked.trackAdding(AbstractTracked.java:256)
    at org.osgi.util.tracker.AbstractTracked.track(AbstractTracked.java:229)
    at org.osgi.util.tracker.ServiceTracker$Tracked.serviceChanged(ServiceTracker.java:901)
    at org.eclipse.osgi.internal.serviceregistry.FilteredServiceListener.serviceChanged(FilteredServiceListener.java:109)
    at org.eclipse.osgi.internal.framework.BundleContextImpl.dispatchEvent(BundleContextImpl.java:914)
    at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230)
    at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:148)
    at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEventPrivileged(ServiceRegistry.java:862)
    at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEvent(ServiceRegistry.java:801)
    at org.eclipse.osgi.internal.serviceregistry.ServiceRegistrationImpl.register(ServiceRegistrationImpl.java:127)
    at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.registerService(ServiceRegistry.java:225)
    at org.eclipse.osgi.internal.framework.BundleContextImpl.registerService(BundleContextImpl.java:464)
    at org.eclipse.osgi.internal.framework.BundleContextImpl.registerService(BundleContextImpl.java:482)
    at org.eclipse.smarthome.automation.internal.core.provider.ModuleTypeResourceBundleProvider.importData(ModuleTypeResourceBundleProvider.java:246)
    at org.eclipse.smarthome.automation.internal.core.provider.AbstractResourceBundleProvider.processAutomationProvider(AbstractResourceBundleProvider.java:317)
    at org.eclipse.smarthome.automation.internal.core.provider.AutomationResourceBundlesEventQueue.processBundleChanged(AutomationResourceBundlesEventQueue.java:313)
    at org.eclipse.smarthome.automation.internal.core.provider.AutomationResourceBundlesEventQueue.run(AutomationResourceBundlesEventQueue.java:144)
    at java.lang.Thread.run(Thread.java:744) 
```

**Comment by marinmitev** [comment]
The last issue found was that on slow machines the tests are executed before parsing of json resources is completed and rules may be in state NOT_INITIALIZED during the test execution, the tests now wait for the rule to become IDLE which means that json files for their modules are already parsed

**Comment by BenediktNiehues** [comment]
Hi Marin, I added some more outputs to the test on my side to find out what is happening exactly: 
Somehow it turns out, that the GenericEventTriggerHandler which is instantiated for the Rule will be disposed during the test-execution and it is not re-created again, although the rule is set to IDLE later. This causes that the EventListener is unregistered and the rule is not executed anymore.

```
Running org.eclipse.smarthome.automation.integration.test.AutomationIntegrationJsonTest
Registered EventSubscriber for Topic: smarthome/items/* and Module: ItemStateChangeTriggerID
java.lang.IllegalArgumentException: Condition Type "EventCondition" does not exist!
    at org.eclipse.smarthome.automation.core.util.ConnectionValidator.validateConditionConnections(ConnectionValidator.java:201)
    at org.eclipse.smarthome.automation.core.util.ConnectionValidator.validateConnections(ConnectionValidator.java:73)
    at org.eclipse.smarthome.automation.core.util.ConnectionValidator.validateConnections(ConnectionValidator.java:56)
    at org.eclipse.smarthome.automation.internal.core.RuleEngine.setRule(RuleEngine.java:316)
    at org.eclipse.smarthome.automation.internal.core.RuleRegistryImpl.addIntoRuleEngine(RuleRegistryImpl.java:87)
    at org.eclipse.smarthome.automation.internal.core.RuleRegistryImpl.add(RuleRegistryImpl.java:77)
    at org.eclipse.smarthome.automation.internal.core.RuleRegistryImpl.add(RuleRegistryImpl.java:1)
    at org.eclipse.smarthome.automation.internal.core.provider.RuleResourceBundleImporter.importData(RuleResourceBundleImporter.java:209)
    at org.eclipse.smarthome.automation.internal.core.provider.RuleResourceBundleImporter.processAutomationProvider(RuleResourceBundleImporter.java:183)
    at org.eclipse.smarthome.automation.internal.core.provider.AutomationResourceBundlesEventQueue.processBundleChanged(AutomationResourceBundlesEventQueue.java:319)
    at org.eclipse.smarthome.automation.internal.core.provider.AutomationResourceBundlesEventQueue.run(AutomationResourceBundlesEventQueue.java:144)
    at java.lang.Thread.run(Thread.java:744)
disposing GenericEventTriggerHandler for moduleID: ItemStateChangeTriggerID
java.lang.IllegalArgumentException: Condition Type "EventCondition" does not exist!
    at org.eclipse.smarthome.automation.core.util.ConnectionValidator.validateConditionConnections(ConnectionValidator.java:201)
    at org.eclipse.smarthome.automation.core.util.ConnectionValidator.validateConnections(ConnectionValidator.java:73)
    at org.eclipse.smarthome.automation.core.util.ConnectionValidator.validateConnections(ConnectionValidator.java:56)
    at org.eclipse.smarthome.automation.internal.core.RuleEngine.setRule(RuleEngine.java:316)
    at org.eclipse.smarthome.automation.internal.core.RuleEngine.moduleTypeUpdated(RuleEngine.java:1073)
    at org.eclipse.smarthome.automation.internal.core.type.ModuleTypeManager.addingService(ModuleTypeManager.java:217)
    at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:941)
    at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:1)
    at org.osgi.util.tracker.AbstractTracked.trackAdding(AbstractTracked.java:256)
    at org.osgi.util.tracker.AbstractTracked.track(AbstractTracked.java:229)
    at org.osgi.util.tracker.ServiceTracker$Tracked.serviceChanged(ServiceTracker.java:901)
    at org.eclipse.osgi.internal.serviceregistry.FilteredServiceListener.serviceChanged(FilteredServiceListener.java:109)
    at org.eclipse.osgi.internal.framework.BundleContextImpl.dispatchEvent(BundleContextImpl.java:914)
    at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230)
    at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:148)
    at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEventPrivileged(ServiceRegistry.java:862)
    at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEvent(ServiceRegistry.java:801)
    at org.eclipse.osgi.internal.serviceregistry.ServiceRegistrationImpl.register(ServiceRegistrationImpl.java:127)
    at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.registerService(ServiceRegistry.java:225)
    at org.eclipse.osgi.internal.framework.BundleContextImpl.registerService(BundleContextImpl.java:464)
    at org.eclipse.osgi.internal.framework.BundleContextImpl.registerService(BundleContextImpl.java:482)
    at org.eclipse.smarthome.automation.internal.core.provider.ModuleTypeResourceBundleProvider.importData(ModuleTypeResourceBundleProvider.java:246)
    at org.eclipse.smarthome.automation.internal.core.provider.AbstractResourceBundleProvider.processAutomationProvider(AbstractResourceBundleProvider.java:317)
    at org.eclipse.smarthome.automation.internal.core.provider.AutomationResourceBundlesEventQueue.processBundleChanged(AutomationResourceBundlesEventQueue.java:313)
    at org.eclipse.smarthome.automation.internal.core.provider.AutomationResourceBundlesEventQueue.run(AutomationResourceBundlesEventQueue.java:144)
    at java.lang.Thread.run(Thread.java:744)
disposing GenericEventTriggerHandler for moduleID: ItemStateChangeTriggerID
RuleEvent: smarthome/rules/org_eclipse_smarthome_automation_integration_test_0/state --> {"status":"NOT_INITIALIZED","statusDetail":"HANDLER_INITIALIZING_ERROR"}
RuleEvent: smarthome/rules/org_eclipse_smarthome_automation_integration_test_0/state --> {"status":"IDLE","statusDetail":"NONE"}
AutoUpdate Item -> myPresenceItem
AutoUpdate Item -> myMotionItem
```

**Comment by BenediktNiehues** [comment]
The test failed because the ruleEngine called directly the dispose method of the Handlers. But the Handler itself was cached in the handlerFactory and reused when again getting the handler. This is solved by PR#103
